stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -1271,5 +1280,265 @@ const Metadata *swift::findConformingSuperclass(
+///
+/// \param imageAddress The address of a Mach header or of a handle returned
+///   from \c dlopen().
+/// \param maybeDlopenHandle Whether or not \a imageAddress might be a
+///   \c dlopen() handle. If it is, an additional call to dyld is made to get
+///   its corresponding Mach header.
+///
+/// If the image does not have a conformance section, returns \c llvm::None.
+static llvm::Optional<ConformanceSection>
+_getConformanceSectionFromMachImage(const void *imageAddress,
+                                    bool maybeDlopenHandle) {
+#ifdef __LP64__
+  typedef const mach_header_64 *MachHeaderAddress;
+#else
+  typedef const mach_header *MachHeaderAddress;
+#endif
+  auto machHeader = reinterpret_cast<MachHeaderAddress>(imageAddress);
+
+  if (maybeDlopenHandle) {
+    // The input pointer might be a dlopen() handle instead of a Mach header.
+    // Try to get the corresponding Mach header.
+    auto dyldAddress = dlsym(const_cast<void *>(imageAddress), "__dso_handle");
+    if (dyldAddress) {
+      machHeader = reinterpret_cast<MachHeaderAddress>(dyldAddress);
+    }
+
+    // If dyld failed to get a Mach header from the image address, it might mean
+    // it's already a Mach header or that it could not resolve the Mac header.
+    // getsectiondata() does not check that the header passed to it is valid, so
+    // perform a basic validation before passing along the pointer we have.
+    if (machHeader->magic != MH_MAGIC && machHeader->magic != MH_MAGIC_64) {
+      return llvm::None;
+    }
+  }
+  
+  unsigned long sectionSize = 0;
+  const auto& sectionData = getsectiondata(machHeader,
+                                           SEG_TEXT,
+                                           MachOProtocolConformancesSection,
+                                           &sectionSize);
+  if (sectionData) {
+    return ConformanceSection(sectionData, sectionSize);
+  }
+
+  return llvm::None;
+}
+#endif
+
+SWIFT_CC(swift)
+static void _enumerateConformingTypesFromImage(
+  const void *imageAddress,
+  const llvm::ArrayRef<ProtocolDescriptorRef>& protocols,
+  TypeEnumerationFunction body,
+  void *bodyContext,
+  SWIFT_CONTEXT void *context,
+  SWIFT_ERROR_RESULT SwiftError **error) {
+
+  bool stop = false;
+  *error = nullptr;
+
+#if defined(__MACH__)
+  // This platform has dyld and we can use it instead of walking the section
+  // table. Takes care of the nuances of DYLD shared cache support.
+  auto section = _getConformanceSectionFromMachImage(imageAddress, true);
+  if (section) {
+    _enumerateConformingTypesFromSections(section.getPointer(), 1,
+                                          protocols, body, bodyContext,
+                                          &stop, context, error);
+  }
+
+#else
+  // This platform doesn't have dyld, so walk all sections in the conformances
+  // table and try to find one that matches the input image address.
+  auto snapshot = Conformances.get().SectionsToScan.snapshot();
```

Comment:
The image registrar should give you the pointer and size of the protocol conformance section directly for a particular image.  The problem is mapping from the image base address to the image registration table.  However, that is something which the runtime has knowledge of - it has all the data for each module.

Reply:
At the time an image is loaded, the image address is certainly available (or, at least, resolvable with a small amount of elbow grease.) But the signature for `swift_addNewDSOImage()` does not have an `imageAddress` argument. Adding that argument would be ABI-breaking on ABI-stable platforms. If we only added it on ABI-unstable platforms, we'd end up bifurcating the relevant code in some pretty unfortunate ways.

Reply:
Actually, looking closer, I think that's a blatant lie and I _can_ get this value, at least for Mach-O/ELF/COFF binaries.